overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+	if email == "" {
+		return CreatedUser{}, &UserError{Err: fmt.Errorf("cannot create user: 'email' field is empty")}
+	}
+
+	db := assertstate.DB(st)
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  nil,
+		serialAs: nil,
+		isSudoer: sudoer,
+	}
+
+	storeService := snapstate.Store(st, nil)
+	st.Unlock()
+	username, opts, err := getUserDetailsFromStore(storeService, email)
+	st.Lock()
+
+	if err != nil {
+		return CreatedUser{}, &UserError{Err: err}
+	}
+	opts.Sudoer = sudoer
+	createdUser, err := u.createUser(st, opts, username, email)
+	if err != nil {
+		return CreatedUser{}, err
+	}
+	return createdUser, nil
+}
+
+// CreateKnownUsers create known user(s)
+// user details are fetched from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If email is passed, only corresponding system user assertion is used for user creation.
+func CreateKnownUsers(st *state.State, sudoer bool, email string) ([]CreatedUser, error) {
+	model, err := findModel(st)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create user: cannot get model assertion: %v", err)
+	}
+
+	serial, err := findSerial(st, nil)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, fmt.Errorf("cannot create user: cannot get serial: %v", err)
+	}
+
+	db := assertstate.DB(st)
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	if email == "" {
+		return u.createAllKnownSystemUsers(st)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+	if err != nil {
+		return nil, &UserError{Err: err}
+	}
+
+	opts.Sudoer = sudoer
+	createdUser, err := u.createUser(st, opts, username, email)
+	if err != nil {
+		return nil, err
+	}
+	return []CreatedUser{createdUser}, nil
+}
+
+// RemoveUser removes linux user account of passed username.
+func RemoveUser(st *state.State, username string) (*auth.UserState, error) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Err: fmt.Errorf("need a username to remove")}
+	}
+
+	// check the user is known to snapd
+	_, err := auth.UserByUsername(st, username)
+	if err == auth.ErrInvalidUser {
+		return nil, &UserError{Err: fmt.Errorf("user %q is not known", username)}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// first remove the system user
+	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
+		return nil, err
+	}
+
+	// then the UserState
+	u, err := auth.RemoveUserByUsername(st, username)
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, err
+	}
+	return u, nil
+}
+
+func (u *createUserOpts) createUser(st *state.State, opts *osutil.AddUserOptions, username, email string) (CreatedUser, error) {
```

Comment:
this needs to move down directly after the struct definition

Reply:
Fixed, nice catch!